src/header/common/upgrade.rs

```diff
@@ -34,43 +32,93 @@ header! {
         test_header!(
             test1,
             vec![b"HTTP/2.0, SHTTP/1.3, IRC/6.9, RTA/x11"],
-            Some(HeaderField(vec![
-                Protocol::ProtocolExt("HTTP/2.0".to_string()),
-                Protocol::ProtocolExt("SHTTP/1.3".to_string()),
-                Protocol::ProtocolExt("IRC/6.9".to_string()),
-                Protocol::ProtocolExt("RTA/x11".to_string()),
+            Some(Upgrade(vec![
+                Protocol::new(ProtocolName::Http, Some("2.0".to_string())),
+                Protocol::new(ProtocolName::Unregistered("SHTTP".to_string()), Some("1.3".to_string())),
+                Protocol::new(ProtocolName::Unregistered("IRC".to_string()), Some("6.9".to_string())),
+                Protocol::new(ProtocolName::Unregistered("RTA".to_string()), Some("x11".to_string())),
                 ])));
     }
 }
 
-/// Protocol values that can appear in the Upgrade header.
-// TODO: Parse version part seperately
-#[derive(Clone, PartialEq, Debug)]
-pub enum Protocol {
-    /// The websocket protocol.
+/// A protocol name used to identify a spefic protocol. Names are case-sensitive
+/// except for the `WebSocket` value.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum ProtocolName {
+    /// `HTTP` value, Hypertext Transfer Protocol
+    Http,
+    /// `TLS` value, Transport Layer Security [RFC2817](http://tools.ietf.org/html/rfc2817)
+    Tls,
+    /// `WebSocket` value, matched case insensitively,Web Socket Protocol
+    /// [RFC6455](http://tools.ietf.org/html/rfc6455)
     WebSocket,
-    /// Some other less common protocol.
-    ProtocolExt(String),
+    /// `h2c` value, HTTP/2 over cleartext TCP
+    H2c,
+    /// Any other protocol name not known to hyper
+    Unregistered(String),
 }
 
-impl FromStr for Protocol {
+impl FromStr for ProtocolName {
     type Err = ();
+    fn from_str(s: &str) -> Result<ProtocolName, ()> {
+        Ok(match s {
+            "HTTP" => ProtocolName::Http,
+            "TLS" => ProtocolName::Tls,
+            "h2c" => ProtocolName::H2c,
+            _ => {
+                if let UniCase("websocket") = UniCase(s) {
+                    ProtocolName::WebSocket
+                } else {
+                    ProtocolName::Unregistered(s.to_string())
+                }
+            }
+        })
+    }
+}
+
+impl Display for ProtocolName {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(match *self {
```

Comment:
Technically could be f.write_str... But might not be consistent with elsewhere in hyper. 

Might be worth a separate issue to change those, as it is a little faster.